RFC - Implement a PSR16 bridge for existing Cache adapters

markstory opened the pull request:
I'd like to get some feedback on the approach taken here before completing all the work. This set of changes takes a different approach to get PSR16 compliant than #11848.

As detailed in #12483 this delivers the new bridge adapter and the `Cache::pool()` method. I've not yet completed the internal refactoring of `Cache` but will do that if folks are on board with the general direction.

markstory committed:
Add new dependency for psr16 interfaces.

markstory committed:
Implement several of the methods needed for psr16

Implement several psr16 methods and many of the tests.

markstory committed:
Complete first draft of simple cache adapter implementation.

markstory committed:
Add pool() method to Cache

This method exposes the new adapter layer to applications. The internals
of Cache have not been updated yet as I want to get feedback on the
approach first.

stickler-ci committed:
Fixing style errors.

markstory committed:
Add missing doc strings.

lorenzo commented:
I like this approach!

I wonder if we could add a boolean property to engines to remember that the last fetch was a failure, instead of relying on the `false` return value. That would allow anyone to cache `false` as the result of a computation.

markstory commented:
@lorenzo Storing false could be enabled if we change the miss value to `null`. This is something I'd like to do in 4.x.

My plan for these changes in 4.x is to drop out interface and replace it with the PSR16 one. I am thinking of keeping `Cache\CacheEngine` with its current set of methods to help with upgrading. `CacheEngine` would use the wrapper logic used here to provide compatibility. The adapters we provide with CakePHP would only implement the PSR16 interface.

ravage84 reviewed (commented):
Just some thoughts and the usual nitpicking. If you agree, I can fix 'em.

markstory commented on src/Cache/Cache.php:
```diff
@@ -234,6 +234,17 @@ public static function engine($config)
+     * @param string $config The name of the configured cache backend.
+     * @return \Cake\Cache\SimpleCacheEngine
+     */
+    public static function pool($config)
```
There is no equivalent setter, so I didn't see a point in having a get prefix.

ravage84 replied:
As you may know, I'm not really up to the speed when it comes to our recent developments. I remembered we went away from indistinguishable and double function getter/setters-methods. So I thought, we always make it clear through the name what it is...

markstory replied:
Yes we stopped adding combined get/set methods as adding typehints would have been impractical.  Those combined methods were split into get/setFoo methods. However, if we only have a getter we have not thus far been adding the get prefix.

markstory commented on src/Cache/SimpleCacheEngine.php:
```diff
@@ -0,0 +1,220 @@
+     *
+     * @param \Cake\Cache\CacheEngine
+     */
+    protected $inner;
```
I'm not too fussy. I chose a short name because I'm a slow typer.

markstory commented on tests/TestCase/Cache/SimpleCacheEngineTest.php:
```diff
@@ -0,0 +1,299 @@
+        $this->cache->setMultiple($data);
+
+        $results = $this->cache->getMultiple(array_keys($data));
+        $this->assertSame($data, $results);
```
The spec doesn't make any guarantees on key order.

markstory commented:
@ravage84 Your suggestions sound good to me. Go ahead with them if you have time/energy. I can get to the changes later this week if you don't get there first.

ravage84 committed:
Rename imported InvalidInterface to InvalidArgumentInterface

ravage84 committed:
Add link to PSR-16 specs

ravage84 committed:
Rename $inner to $innerEngine

ravage84 committed:
Typo

ravage84 committed:
Rename checkKey() to ensureValidKey()

ravage84 committed:
Reword invalid cache key exception slightly

ravage84 committed:
Reword method doc block

ravage84 committed:
Reword exception explanation

ravage84 committed:
Reword method short description

ravage84 committed:
Reword exception explanation

ravage84 committed:
Reword exception explanation

ravage84 committed:
Reword exception explanation

ravage84 committed:
Reword exception explanation

ravage84 committed:
Reword exception explanation

ravage84 committed:
Add ensureValidKeys() to ensure the validity of cache key sets

This satisfies the requirements of getMultiple() & deleteMultiple() in the PSR-16 specs:

> "MUST be thrown if $keys is neither an array nor a Traversable,or if any of the $keys are not a legal value."

But still does not fully satisfy the requirements of setMultiple(), as currently only the keys are validated:

> "MUST be thrown if $values is neither an array nor a Traversable,or if any of the $values are not a legal value."

ravage84 committed:
Add test coverage annotation to improve the meaningfulness of the coverage

ravage84 committed:
Start with capital letter

ravage84 committed:
Not only expect the exception but also its message

ravage84 committed:
Add property type hints

Improves auto-completion & IDE auto resolution.

ravage84 committed:
Rename $inner to $innerEngine to use the same name as in the subject under test

ravage84 committed:
Improve test readability

ravage84 committed:
Improve test readability

ravage84 committed:
Loosen test assertion as the specs do not guarantee the key order

ravage84 committed:
TTL, not ttl

ravage84 committed:
Fix CS

ravage84 committed:
Fix CS

ravage84 commented:
@markstory I did my changes on the PSR-16 wrapper. Feel free to revert anything if it does not adhere to our curent (quality) standards.

Please have a special look at b1cce65c7cbb715a8f599a0b0a2655105216d074. We still do not fully satisfy the specs as we do not check for the value validity. I did not want to implement that without getting back at you, first.

markstory commented:
@ravage84 I don't think checking the values is as important as checking the keys. If a value cannot be stored the cache backend will fail.

markstory committed:
Fix phpcs and phpstan errors.

markstory commented:
Any other feedback on this?
